HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc

```diff
@@ -0,0 +1,255 @@
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <random>
+
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h"
+
+using namespace cms::alpakatools;
+using namespace ALPAKA_ACCELERATOR_NAMESPACE;
+
+static constexpr auto s_tag = "[" ALPAKA_TYPE_ALIAS_NAME(alpakaTestHistoContainer) "]";
+
+template <typename T, typename Hist, typename VERIFY, typename INCR>
+void checkContents(Hist* h,
+                   unsigned int N,
+                   VERIFY& verify,
+                   INCR& incr,
+                   T window,
+                   uint32_t nParts,
+                   const T* v,
+                   const uint32_t* offsets) {
+  for (uint32_t j = 0; j < nParts; ++j) {
+    auto off = Hist::histOff(j);
+    for (uint32_t i = 0; i < Hist::nbins(); ++i) {
+      auto ii = i + off;
+      if (0 == h->size(ii))
+        continue;
+      auto k = *h->begin(ii);
+      if (j % 2)
+        k = *(h->begin(ii) + (h->end(ii) - h->begin(ii)) / 2);
+#ifndef NDEBUG
+      auto bk = h->bin(v[k]);
+#endif
+      ALPAKA_ASSERT_OFFLOAD(bk == i);
```

Comment:
It is not clear to me if in the current setup `ALPAKA_ASSERT_OFFLOAD` actually expands to an assertion (or similar) on all backends. I see @fwyzard wondered this out loud in https://github.com/alpaka-group/alpaka/issues/2001 . In addition, in the `CUDAUtilities` all the tests are compiled with `-DGPU_DEBUG` to make sure the assertions are done also on GPU (in conjunction with
https://github.com/cms-sw/cmssw/blob/e4bfeac8c9a8d7e1690b8a16cd8464e0330b24dd/HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h#L4-L16
)

Reply:
Agreed... maybe we can converge on what the _desired_ behaviour for `ALPAKA_ASSERT_OFFLOAD` should be, and then we can propose it to the Alpaka developers, or implement our own variant.

Reply:
See https://github.com/alpaka-group/alpaka/issues/2186 .